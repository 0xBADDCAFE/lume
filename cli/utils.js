import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
