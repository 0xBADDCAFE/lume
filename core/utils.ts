import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import {
  brightGreen,
  brightYellow,
  cyan,
  dim,
  green,
  red,
} from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";
import { Exception } from "./errors.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "netlify_cms",
  "on_demand",
  "pagefind",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

/** A list of the available plugins with init configurations */
export const initPlugins = [
  "jsx",
  "jsx_preact",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.25.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  let global = false;
  try {
    await Promise.any([
      Deno.stat("deno.json"),
      Deno.stat("deno.jsonc"),
    ]);
    await Deno.stat("import_map.json");
  } catch {
    global = true;
  }

  const command = global
    ? (stable ? "lume upgrade --global" : "lume upgrade --dev --global")
    : (stable ? "deno task lume upgrade" : "deno task lume upgrade --dev");

  log(
    `Update available ${dim(current)}  â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
// deno-lint-ignore ban-types
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string) {
  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      return path;
    }
  }

  return posix.join("/", path);
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch (cause) {
      throw new Exception("Config file not found", { path, cause });
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(
  importMapFile?: string,
): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      importMapFile ||= config.importMap;
      if (importMapFile) {
        config.importMap = importMapFile;
        const importMap: ImportMap = isUrl(importMapFile)
          ? await (await fetch(importMapFile)).json()
          : await JSON.parse(await Deno.readTextFile(importMapFile));

        return { file, config, importMap };
      }
      return { file, config };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.importMap = "./import_map.json";
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    console.log(importMapFile);

    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    console.log(
      brightYellow("deno.jsonc needs to be manually updated:"),
      "Use deno.json to update it automatically",
      JSON.stringify(config, null, 2),
    );
  } else {
    await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
    console.log(brightGreen("Deno configuration file saved:"), file);
  }
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}

export function createDate(str: string | number): Date | undefined {
  if (typeof str === "number") {
    return new Date(str);
  }

  const datetime = str.match(
    /^(\d{4})-(\d\d)-(\d\d)(?:-(\d\d)-(\d\d)(?:-(\d\d))?)?$/,
  );

  if (datetime) {
    const [, year, month, day, hour, minute, second] = datetime;

    return new Date(Date.UTC(
      parseInt(year),
      parseInt(month) - 1,
      parseInt(day),
      hour ? parseInt(hour) : 0,
      minute ? parseInt(minute) : 0,
      second ? parseInt(second) : 0,
    ));
  }

  if (str.match(/^\d+$/)) {
    return new Date(parseInt(str));
  }
}

export async function read(path: string, isBinary: true): Promise<Uint8Array>;
export async function read(path: string, isBinary: false): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const response = await fetch(path);

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}
