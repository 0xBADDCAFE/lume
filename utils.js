import { bold, red } from "./deps/colors.js";
import { SEP } from "./deps/path.js";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".html", "text/html; charset=utf-8"],
  [".txt", "text/plain; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".css", "text/css; charset=utf-8"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".json", "application/json"],
  [".webmanifest", "application/manifest+json"],
  [".ico", "image/x-icon"],
  [".avif", "image/avif"],
  [".png", "image/png"],
  [".apng", "image/apng"],
  [".jpg", "image/jpg"],
  [".jpeg", "image/jpg"],
  [".gif", "image/gif"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mpeg"],
  [".xml", "text/xml"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".wasm", "application/wasm"],
  [".webp", "image/webp"],
  [".webm", "video/webm"],
  [".zip", "application/zip"],
]);

export function error(context, message, exception) {
  console.error(bold(red(`${context}:`)), message);
  if (exception) {
    console.error(exception);
  }
  console.error("");
}

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (
      typeof merged[key] === "object" && typeof value === "object" &&
      !Array.isArray(merged[key])
    ) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

export function normalizePath(path) {
  // Is Windows path
  if (SEP !== "/") {
    return path.replaceAll(SEP, "/");
  }

  return path;
}

export const slugChars = new Map([
  ["รฐ", "d"],
  ["ร", "ss"],
]);

export function slugify(string) {
  return string.toLowerCase()
    .normalize("NFKD")
    .replaceAll(/[\s_-]+/g, "-")
    .replaceAll(/[^\w\/-]/g, (char) => slugChars.get(char) || "");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}
