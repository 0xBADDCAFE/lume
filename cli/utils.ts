import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
