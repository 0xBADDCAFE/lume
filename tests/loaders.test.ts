import { assert, assertStrictEquals as equals } from "../deps/assert.ts";
import binaryLoader from "../core/loaders/binary.ts";
import { getPage, getSite, testPage } from "./utils.ts";

Deno.test("load the pages of a site", async () => {
  const site = getSite({
    dev: true,
    src: "normal",
  });

  site.loadAssets([".png"], binaryLoader);
  site.copy("static.yml");
  site.loadAssets([".css"]);

  await site.build();

  // Test the generated pages
  equals(site.pages.length, 8);

  // The data is merged
  testPage(site, "/pages/1_page1", (page) => {
    const metas = page.data.metas as Record<string, unknown>;
    const imagick = page.data.imagick as Record<string, unknown>;

    assert(page.data.draft);
    equals(page.data.url, "/pages/page1/");
    equals(page.data.title, "Page 1");
    equals(page.data.tags?.length, 3);
    equals(page.data.tags?.[0], "pages");
    equals(page.data.tags?.[1], "sub-pages");
    equals(page.data.tags?.[2], "page1");
    equals(metas.title, "Custom title");
    equals(metas.description, "Default description");
    equals(imagick.length, 2);
    equals(page.data.site, "Folder overrided site name");
    equals(page.dest.path, "/pages/page1/index");
    equals(page.dest.ext, ".html");
    equals(page.data.date?.getTime(), 1);

    // Shared data
    // @ts-ignore: unknown property
    equals(page.data.colors.length, 3);
    // @ts-ignore: unknown property
    equals(page.data.documents.length, 3);
    // @ts-ignore: unknown property
    equals(page.data.drinks.alcoholic.length, 2);
    // @ts-ignore: unknown property
    equals(page.data.names.length, 2);
  });

  testPage(site, "/pages/2020-06-21_page2", (page) => {
    assert(!page.data.draft);
    equals(page.data.url, "/overrided-page2/");
    equals(page.data.title, "Page 2");
    equals(page.data.tags?.length, 2);
    equals(page.data.tags?.[0], "pages");
    equals(page.data.tags?.[1], "sub-pages");
    equals(page.data.site, "Folder overrided site name");
    equals(page.dest.path, "/overrided-page2/index");
    equals(page.dest.ext, ".html");
    equals(
      page.data.date?.getTime(),
      new Date(Date.UTC(2020, 5, 21)).getTime(),
    );
  });

  // Shared data
  const page1 = getPage(site, "/pages/1_page1");
  const page2 = getPage(site, "/pages/2020-06-21_page2");

  // @ts-ignore: unknown property
  assert(page1.data.colors === page2.data.colors);
  // @ts-ignore: unknown property
  assert(page1.data.documents === page2.data.documents);
  // @ts-ignore: unknown property
  assert(page1.data.drinks === page2.data.drinks);
  // @ts-ignore: unknown property
  assert(page1.data.names === page2.data.names);

  testPage(site, "/pages/page3", (page) => {
    assert(!page.data.draft);
    equals(page.data.url, "/page_3");
    equals(page.data.title, "Page 3");
    equals(page.data.tags?.length, 2);
    equals(page.data.tags?.[0], "pages");
    equals(page.data.tags?.[1], "sub-pages");
    equals(page.data.site, "Folder overrided site name");
    equals(page.dest.path, "/page_3");
    equals(page.dest.ext, "");
    equals(page.data.date?.getTime(), new Date(2020, 0, 1).getTime());
  });

  testPage(site, "/pages/2021-01-02-18-32_page4", (page) => {
    assert(!page.data.draft);
    equals(page.data.url, "/pages/page4/");
    equals(page.data.title, "Page 4");
    equals(page.data.tags?.length, 2);
    equals(page.data.tags?.[0], "pages");
    equals(page.data.tags?.[1], "sub-pages");
    equals(page.data.site, "Overrided site name");
    equals(page.dest.path, "/pages/page4/index");
    equals(page.dest.ext, ".html");
    equals(
      page.data.date?.getTime(),
      new Date(Date.UTC(2021, 0, 2, 18, 32)).getTime(),
    );
  });

  testPage(site, "/page5", (page) => {
    assert(!page.data.colors);
    assert(!page.data.documents);
    assert(!page.data.draft);
    equals(page.data.url, "/page5/");
    equals(page.data.title, "Page 5");
    equals(page.data.tags?.length, 1);
    equals(page.data.tags?.[0], "pages");
    equals(page.data.site, "Default site name");
    equals(page.dest.path, "/page5/index");
    equals(page.dest.ext, ".html");
    equals(
      page.data.date?.getTime(),
      new Date(Date.UTC(1979, 5, 21, 23, 45, 0)).getTime(),
    );
  });

  testPage(site, "/pages/2021-12-29_page6", (page) => {
    equals(page.data.url, "/pages/page6/");
    equals(page.data.title, "Page 6");
    equals(page.data.tags?.length, 3);
    equals(page.data.tags?.[0], "pages");
    equals(page.data.site, "Folder overrided site name");
    equals(page.dest.path, "/pages/page6/index");
    equals(page.dest.ext, ".html");
    equals(
      page.data.date?.getTime(),
      new Date(Date.UTC(2022, 0, 1, 0, 0, 0)).getTime(),
    );
  });

  // Test binary loader
  testPage(site, "/favicon", (page) => {
    equals(page.data.url, "/favicon.png");
    equals(page.dest.path, "/favicon");
    equals(page.dest.ext, ".png");
    equals(typeof page.content, "object");
    assert(page.content instanceof Uint8Array);
  });

  // Test non HTML pages
  testPage(site, "/styles", (page) => {
    assert(page.document === undefined);
    equals(page.data.url, "/styles.css");
    equals(page.dest.path, "/styles");
    equals(page.dest.ext, ".css");
    equals(typeof page.content, "string");
  });
});

Deno.test("ignored draft pages on dev=false", async () => {
  const site = getSite({
    dev: false,
    src: "normal",
  });

  site.copy("static.yml");

  await site.build();

  equals(site.pages.length, 5);
});
