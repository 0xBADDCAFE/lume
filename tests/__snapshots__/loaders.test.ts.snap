export const snapshot = {};

snapshot[`Load the pages of a site 1`] = `9`;

snapshot[`Load the pages of a site 2`] = `
{
  formats: [
    {
      includesPath: undefined,
      key: ".tmpl.ts",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".tmpl.js",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".tmpl.json",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".json",
      pageType: undefined,
    },
    {
      includesPath: undefined,
      key: ".md",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".js",
      pageType: undefined,
    },
    {
      includesPath: undefined,
      key: ".ts",
      pageType: undefined,
    },
    {
      includesPath: "_includes",
      key: ".njk",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".yaml",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".yml",
      pageType: "page",
    },
    {
      includesPath: undefined,
      key: ".png",
      pageType: "asset",
    },
    {
      includesPath: undefined,
      key: ".css",
      pageType: "asset",
    },
  ],
}
`;

snapshot[`Load the pages of a site 3`] = `
[
  {
    dest: "/static.yml",
    saved: true,
    src: "/static.yml",
  },
]
`;

snapshot[`Load the pages of a site 4`] = `
{
  content: "Uint8Array(1503)",
  data: {
    content: "Uint8Array(1503)",
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Default site name",
    tags: [
      "pages",
    ],
    url: "/favicon.png",
  },
  dest: {
    ext: ".png",
    path: "/favicon",
  },
  src: {
    ext: ".png",
    path: "/favicon",
  },
}
`;

snapshot[`Load the pages of a site 5`] = `
{
  content: "Content of Page 5",
  data: {
    content: "Content of Page 5",
    date: 1970-01-01T00:00:00.000Z,
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Default site name",
    tags: [
      "pages",
    ],
    title: "Page 5",
    url: "/page5/",
  },
  dest: {
    ext: ".html",
    path: "/page5/index",
  },
  src: {
    ext: ".yaml",
    path: "/page5",
  },
}
`;

snapshot[`Load the pages of a site 6`] = `
{
  content: "<p>Content of Page 1</p>
",
  data: {
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "
Content of Page 1
",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    draft: true,
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
      {
        resize: [
          900,
          900,
        ],
        suffix: "-big",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Custom title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "pages",
      "sub-pages",
      "page1",
    ],
    title: "Page 1",
    url: "/pages/page1/",
  },
  dest: {
    ext: ".html",
    path: "/pages/page1/index",
  },
  src: {
    ext: ".md",
    path: "/pages/1_page1",
  },
}
`;

snapshot[`Load the pages of a site 7`] = `
{
  content: "Content of Page 2",
  data: {
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 2",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "pages",
      "sub-pages",
    ],
    title: "Page 2",
    url: "/overrided-page2/",
  },
  dest: {
    ext: ".html",
    path: "/overrided-page2/index",
  },
  src: {
    ext: ".tmpl.json",
    path: "/pages/2020-06-21_page2",
  },
}
`;

snapshot[`Load the pages of a site 8`] = `
{
  content: "Content of Page 4 in Overrided site name",
  data: {
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: undefined,
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Overrided site name",
    tags: [
      "pages",
      "sub-pages",
    ],
    title: "Page 4",
    url: "/pages/page4/",
  },
  dest: {
    ext: ".html",
    path: "/pages/page4/index",
  },
  src: {
    ext: ".tmpl.ts",
    path: "/pages/2021-01-02-18-32_page4",
  },
}
`;

snapshot[`Load the pages of a site 9`] = `
{
  content: "<p>Content of Page 6</p>
",
  data: {
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "
Content of Page 6
",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "pages",
      "sub-pages",
      "page6",
    ],
    title: "Page 6",
    url: "/pages/page6/",
  },
  dest: {
    ext: ".html",
    path: "/pages/page6/index",
  },
  src: {
    ext: ".md",
    path: "/pages/2021-12-29_page6",
  },
}
`;

snapshot[`Load the pages of a site 10`] = `
{
  content: "Content of Page 3",
  data: {
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 3",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "pages",
      "sub-pages",
    ],
    title: "Page 3",
    url: "/page_3",
  },
  dest: {
    ext: "",
    path: "/page_3",
  },
  src: {
    ext: ".tmpl.js",
    path: "/pages/page3",
  },
}
`;

snapshot[`Load the pages of a site 11`] = `
{
  content: "Content of Page 7",
  data: {
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 7",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "pages",
      "sub-pages",
      "sub-sub-pages",
      "page7",
    ],
    title: "Page 7",
    url: "/pages/subpage/page7/",
  },
  dest: {
    ext: ".html",
    path: "/pages/subpage/page7/index",
  },
  src: {
    ext: ".tmpl.js",
    path: "/pages/subpage/page7",
  },
}
`;

snapshot[`Load the pages of a site 12`] = `
{
  content: "body {
  font-family: sans-serif;
  color: black;
}
",
  data: {
    content: "body {
  font-family: sans-serif;
  color: black;
}
",
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    paginate: [GeneratorFunction: paginate],
    search: Search {},
    site: "Default site name",
    tags: [
      "pages",
    ],
    url: "/styles.css",
  },
  dest: {
    ext: ".css",
    path: "/styles",
  },
  src: {
    ext: ".css",
    path: "/styles",
  },
}
`;
